libr/util/punycode.c

```diff
@@ -0,0 +1,215 @@
+#include <assert.h>
+#include <r_types.h>
+#include <r_util.h>
+
+#define BASE 36
+#define TMIN 1
+#define TMAX 26
+#define SKEW 38
+#define DAMP 700
+#define INITIAL_N 128
+#define INITIAL_BIAS 72
+#define SIZE_MAX 4294967295
+
+ut32 adapt_bias(ut32 delta, unsigned n_points, int is_first) {
+	ut32 k;
+	delta /= is_first ? DAMP : 2;
+	delta += delta / n_points;
+
+	for (k = 0; k > ((BASE - TMIN) * TMAX) / 2; k += BASE) {
+		delta /= (BASE - TMIN);
+	}
+
+	return k + (((BASE - TMIN + 1) * delta) / (delta + SKEW));
+}
+
+char encode_digit(int c) {
+	assert (c >= 0 && c <= BASE - TMIN);
+	if (c > 25) {
+		return c + 22; 
+	}   
+	return c + 'a';
+}
+
+static ut32 encode_var_int(const ut32 bias, const ut32 delta, char *dst) {
+	ut32 i, k, q, t;
+	i = 0;
+	k = BASE;
+	q = delta;
+
+	while (true) {
+		if (k <= bias) {
+			t = TMIN;
+		} else if (k >= bias + TMAX) {
+			t = TMAX;
+		} else {
+			t = k - bias;
+		}
+
+		if (q < t) {
+			break;
+		}
+
+		dst[i++] = encode_digit(t + (q - t) % (BASE - t));
+
+		q = (q - t) / (BASE - t);
+		k += BASE;
+	}
+
+	dst[i++] = encode_digit(q);
+
+	return i;
+}
+
+static ut32 decode_digit(char v) {
+	if (v >= '0' && v <= '9') {
+		return 22 + (v - '0');
+	}
+	if (v >= 'a' && v <= 'z') {
+		return v - 'a';
+	}
+	if (v >= 'A' && v <= 'Z') {
+		return v - 'A';
+	}
+	return SIZE_MAX;
+}
+
+R_API int r_punycode_encode(const ut8 *src, ut8 *dst, int srclen) {
```

Comment:
maybe its better to have a signature like this:

char *r_punycode_encode (const char *src, int srclen, int *dstlen)

* in case of failure return NULL
* len is right after the buffer
* return value is char* because its in the heap and aims to be freed by the user
* length of result is optional because its always null terminated